lib/auth/auth_lib.c

```diff
@@ -0,0 +1,342 @@
+	}
+}
+
+
+
+/* ========================= external API ============================ */
+
+
+/**
+ * @see auth_lib.h
+ */
+int auth_lib_init(struct authenticate_conn *auth_conn, enum auth_instance_id instance,
+		  auth_status_cb_t status_func, void *context, struct auth_optional_param *opt_params,
+		  enum auth_flags auth_flags)
+{
+	int err = 0;
+
+	/* check input params */
+	if (status_func == NULL) {
+		LOG_ERR("Error, status function is NULL.");
+		return AUTH_ERROR_INVALID_PARAM;
+	}
+
+	/* check auth flags */
+	if (!auth_lib_checkflags(auth_flags)) {
+		LOG_ERR("Invalid auth flags.");
+		return AUTH_ERROR_INVALID_PARAM;
+	}
+
+	/* init the struct to zero */
+	memset(auth_conn, 0, sizeof(struct authenticate_conn));
+
+	/* setup the status callback */
+	auth_conn->status_cb = status_func;
+	auth_conn->callback_context = context;
+
+	auth_conn->cancel_auth = false;
+	auth_conn->instance = instance;
+
+	/* init the work item used to post authentication status */
+	k_work_init(&auth_conn->auth_status_work, auth_lib_status_work);
+
+	auth_conn->is_client = (auth_flags & AUTH_CONN_CLIENT) ? true : false;
+
+#if defined(CONFIG_AUTH_DTLS)
+
+	if (auth_flags & AUTH_CONN_DTLS_AUTH_METHOD) {
+		/* Set the DTLS authentication thread */
+		auth_conn->auth_func = auth_dtls_thead;
+		{
+			if (opt_params == NULL || opt_params->param_id != AUTH_DTLS_PARAM) {
+				LOG_ERR("Missing certificates for TLS/DTLS authentication.");
+				return AUTH_ERROR_INVALID_PARAM;
+			}
+
+			struct auth_dtls_certs *certs = &opt_params->param_body.dtls_certs;
+
+			// init TLS layer
+			err = auth_init_dtls_method(auth_conn, certs);
+		}
+
+		if (err) {
+			LOG_ERR("Failed to initialize MBed TLS, err: %d", err);
+			return err;
+		}
+	}
+#endif
+
+#if defined(CONFIG_AUTH_CHALLENGE_RESPONSE)
+
+	if (auth_flags & AUTH_CONN_CHALLENGE_AUTH_METHOD) {
+		/* Set the Challenge-Response authentication thread */
+		auth_conn->auth_func = auth_chalresp_thread;
+
+		if ((opt_params != NULL) && (opt_params->param_id == AUTH_CHALRESP_PARAM)) {
+
+			struct auth_challenge_resp *chal_resp = &opt_params->param_body.chal_resp;
+
+			err = auth_init_chalresp_method(auth_conn, chal_resp);
+
+			if (err) {
+				LOG_ERR("Failed to set Challege-Response param, err: %d", err);
+				return err;
+			}
+		}
+	}
+#endif
+
+	/*
+	 * Set auth connect into thread param instance.
+	 *
+	 * @note:  k_sem_give() is called to start the thread, which provides
+	 * a compile_barrier() and for SMP systems should handle any
+	 * necessary memory barriers and/or cache syncing.
+	 * The important point is the auth_conn pointer is set into
+	 * a structure used by another thread (auth_thrd_entry)
+	 */
+	thrd_params[instance].auth_conn = auth_conn;
+
+	return AUTH_SUCCESS;
+}
+
+/**
+ * @see auth_lib.h
+ */
+int auth_lib_deinit(struct authenticate_conn *auth_conn)
+{
+	/* Free any resources, nothing for now, but maybe
+	 * needed in the future */
+	return AUTH_SUCCESS;
```

Comment:
don't we need to free any resource ? What happens with mbedTLS context ?

Reply:
Yes, that's an oversight.  Added code to properly de-init the Challenge-Response and DTLS auth methods.